### PR TITLE
[Subtitles] Improvement to manual position/calibration and others related changes

### PR DESCRIPTION
**Pull request by CastagnaIT**
## Description

Main changes:
- Calibration window now can be used with all _bottom_ subtitle position alignments, **then no top alignments** due to libass limits
- Fix broken subtitle bar in calibration window when overscan has been changed
- Fix wrong subtitle position when overscan is set
- Fix wrong notification label when you move subtitle position manually (e.g. shorcut key CTRL+UP/DOWN)
- Implemented simulated acceleration speed when you move subtitle position (e.g. shorcut key CTRL+UP/DOWN), this allow very easy position without waiting an eternity to move the subtitles for every single pixel... (this implementation can be reused for each others settings)
- Fix improper positioning of subs text within the black bars (related below/above alignments), the wrong position could be influenced due to custom user margins, now have fixed value (if needed an user can still move manually)
- Increased default vertical margin (see screenshots) is now higher but not excessively so, i think now can be more accommodating for more users
- New player builtins/actions `SubtitleShiftUp(save)` `SubtitleShiftDown(save)` read below

### How works the changed functionalities

**Moving the subtitles**
Not so different from Kodi 19, you can move subtitles position as you like (e.g. shorcut key CTRL+UP/DOWN),
but the changed position will be done **as temporary way**, so when you will stop the video, in the next video played the subtitle position will be restored to the _default position value._

How to change the default position value permanently:
1) By using the calibration window
2) By changing the temporary "behaviour" to permanent, by using the new player builtins:
For example create/edit the XML keyboard keymaps then replace or map to a key:
```
<up mod="ctrl">SubtitleShiftUp(save)</up>
<down mod="ctrl">SubtitleShiftDown(save)</down>
```
The optional parameter `(save)` allow to save permanently the subtitle position change.

For the less informed: **subtitles format like ASS/SSA/WebVTT** define their own positions of text in the metadata, therefore these subtitles **cannot be moved by default**. If you wish to do so, you must set `Override subtitles styles` setting to `Positions`.

**Difference between vertical margin and manual (shift) position**
On Kodi 19 this was not existent, but now we have this difference:
- The vertical margin influence the text that can be positioned below or top of screen (also for forced tags cases see screenshot) so you can call as "margins of screen"
- The manual position shift the subtitle text from bottom only, and could also place the text offscreen (but we prevent it)

Then from here we can understand the difference of these settings behaviours, maybe we can think in future to add the vertical margin as proper GUI setting

My thoughts: The "Manual" position alignment setting has no difference from "Bottom" align funtionality, so you might want to think about keeping it as backward compat!? or delete this alignment?

## Motivation and context

After my firsts subtitles rework PR's, i was aware of problems with manual subtitle position,
which required a separate study because of the many use cases to be tested.

Meantime users reported discontent in the forum because it was no longer possible to move subtitles easily except by using the "Manual" alignment, other users report that subtitles are placed again too bottom by default (as with previous Kodi versions)
and need to adjust via xml advanced settings is uncomfortable

i hope this will make happy more users

## How has this been tested?

Switch across all alignments, then tested with shorcut key CTRL+UP/DOWN,
using Calibration window during playback,
ensure that subtitles that make use of margins (WebVTT) are not broken and are influenced only with override position setting

~The PR is ready but i have set as WIP to have more time to test it~

## What is the effect on users?

Allow to use manual position with all bottom alignments positions settings (then except top align positions cases)
Better vertical margins etc...

## Screenshots (if appropriate):
COMPARISON TO VERTICAL MARGIN
BEFORE:
Clipboard01alignBefore
AFTER:
Clipboard01alignNow

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h
Comment (garbear) [inline xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h:121]: Would the term "vertical shift" be more accurate than "position"?

**Comment by garbear** [comment]
I like the use of the MovingSpeed util

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h:121]: this in reality set the subtitle vertical position not the shift difference
i can rename it like `SetSubtitleVerticalPosition` / `GetSubtitleVerticalPosition`

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp
Comment (enen92) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp:3317]: Looking at the renderManager calls in this file I'm unsure this particular usage is correct.  To me subtitle vertical position is something that belongs to the player context. We already have:

```
void CVideoPlayer::SetVideoSettings(CVideoSettings& settings)
{
  m_processInfo->SetVideoSettings(settings);
  m_renderManager.SetVideoSettings(settings);
  m_renderManager.SetDelay(static_cast<int>(settings.m_AudioDelay * 1000.0f));
  ... 
}
```

Can't this be saved in the player object and just passed to the render manager like all the other player settings?

File: xbmc/cores/VideoPlayer/VideoRenderers/DebugRenderer.h
Comment (enen92) [inline xbmc/cores/VideoPlayer/VideoRenderers/DebugRenderer.h:41]: We use the notify/observer to inform the overlayrenderer (since it has to react to any of the settings being monitored) and the settings callback to perform individual actions on individual settings change in this component right?

**Review by enen92 — COMMENTED** [review]
I don't really see anything wrong with the implementation if we take into consideration what was there before. However there's a lot of stuff I am not able to review because we don't have a clear architectural direction anymore. Not sure what the global stance on resolution calls, resolution set, g_application usage, etc really are and that affects the future software direction. In summary this is okay in the current state, not sure if it negatively affects any future refactor

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp:271]: I don't know how GraphicContext class can be refactored to grant a more clean access,
but i can give a comment to the last sentence,
currently `iSubtitles` store the subtitle position on each resolution because
if you change resolution the scale is different then you need different value of subs position pixels
(currently these are stored and automatically restored on res change)

if we think to store this value outside resinfo like fixed value, happen this e.g.:

if you set the position to 100px (from bottom) and you have screen at 1080p
when you switch at 4k the subs will be too much bottom screen.

otherwise i think we should manage the position like do Libass as percentage % instead of pixels then we can store a single %value that can be used across all resolutions

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp:3317]: i have implemented to CVideoSettings try to see

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp:539]: Did you consider thread safety? Looks dangerous doing a whole lot of things here on Notify.

Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp:541]: right added

Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp:542]: It is good practice to keep processing in callbacks to a bare minimum. Deliver data, set a flag, but don't do any calculations, calling functions, etc. The locks might block the render thread too long.

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp
Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp]: @enen92 my vision was to make VideoPlayer indenpendent from the GUI component. If VP was indenpendent from the UI, you could even use multiple instances of VP. That would allow PiP or transcoding a video stream by rendering not to the display but to an encoder.
@CastagnaIT independent from the GUI does also mean not to use settings like globals. Settings are mainly relevant when rendering to the display.

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp]: ok then i have replaced all with a single variable change

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp]: @FernetMenta can you provide an example/link of what you mean for global settings?

Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp]: Doing things like
```
const std::shared_ptr<CSettings> settings = CServiceBroker::GetSettingsComponent()->GetSettings();
```
all over the place is like using globals. Basically this creates a dependency to the UI / settings.

Have a look at AE https://github.com/xbmc/xbmc/blob/master/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp#L2635
It loads all settings into a local variable. The other code in AE doesn't make calls to the settings component but uses the local copy.

Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp]: ok understand the way, but map all settings in a variable as on ActiveAE `m_settings`, in my opinion it is not always a big deal and IMO as is now i find it a fragmentary code implementation

so each class have to:
-store all required settings locally
-keep in sync the local settings values in some way

Pro:
-less calls to GUI setting component but in any case there is dependency to settings component, although (should be) confined to one method
Cons:
-as is implemented now, make more code to maintain (if a setting will be changed you have to rework structs and variables of all customs `m_settings` variables of each class, maybe use common structs that can be maintained on one place?)
-more memory occupied with same data, since each class have to store same settings values

at least for OverlayRenderer no need to store all settings used since needs always fresh data

but if i have to take some ispiration from this i could store locally the settings more used only, so for: GetSubtitleVerticalMargin() / ConvertLibass() methods,
but then i need also to keep in sync these local settings stored, this can be done with "Observable Notify()" callback,
but as you specified in comment below is too much to put in the callback, then i need to delegate this update to the "Render()" it's all an mess of "loop"

**Comment by CastagnaIT** [comment]
what we do with this PR? i have to change something else?

File: xbmc/cores/VideoPlayer/DVDCodecs/Overlay/DVDOverlayCodecSSA.cpp
Comment (enen92) [inline xbmc/cores/VideoPlayer/DVDCodecs/Overlay/DVDOverlayCodecSSA.cpp:135]: Can you add a TODO to move this into SubtitleSettings?

File: xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitleParserSSA.cpp
Comment (enen92) [inline xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitleParserSSA.cpp:43]: same here:
Can you add a TODO to move this into SubtitleSettings?

File: xbmc/settings/SubtitlesSettings.cpp
Comment (enen92) [inline xbmc/settings/SubtitlesSettings.cpp:83]: Can you make `ObservableMessageSubtitleAlignChanged` a bit more generic i.e. not "scoped" to subtitles? Something like ObservableMessagePositioningChanged would do for me. Those messages are supposed to be generic so we need to somehow extend them if we want them scoped per component or make them generic enough to be reused by all

File: xbmc/utils/Observer.h
Comment (enen92) [inline xbmc/utils/Observer.h:25]: see comment above, a comment explaining what it means would be great too

**Review by enen92 — APPROVED** [review]
I've runtime tested this and works as expected, globally it's a nice improvement.
At some point in the future the subtitle related settings must likely need to be moved to the SubtitleSettings class with proper get/set methods to simplify the code and their widespread around the codebase.
I'd say to fix the minors, cleanup the fixup commits and go for it

**Comment by CastagnaIT** [comment]
if no other objections i will merge it tomorrow

File: xbmc/utils/MovingSpeed.h
Comment (lrusak) [inline xbmc/utils/MovingSpeed.h:19]: I might have missed it but this doesn't seem to be used anywhere.

File: xbmc/utils/MovingSpeed.h
Comment (lrusak) [inline xbmc/utils/MovingSpeed.h:36]:
```suggestion
  EventCfg(float acceleration, float maxVelocity, uint32_t resetTimeout) : m_acceleration(acceleration), m_maxVelocity(maxVelocity), m_resetTimeout(resetTimeout) {}
```

File: xbmc/windowing/Resolution.h
Comment (lrusak) [inline xbmc/windowing/Resolution.h:44]:
```suggestion
  bool operator==(const OVERSCAN& other)
```

File: xbmc/windowing/Resolution.h
Comment (lrusak) [inline xbmc/windowing/Resolution.h:48]:
```suggestion
  bool operator!=(const OVERSCAN& other)
```

File: xbmc/video/PlayerController.cpp
Comment (lrusak) [inline xbmc/video/PlayerController.cpp:522]:
```suggestion
    std::string strValue = StringUtils::Format("{:.0f}px", slider->GetFloatValue());
```

**Review by lrusak — COMMENTED** [review]
Don't have the time to look at the functional changes so just left a couple comments.

In general I think this PR could benefit with some added whitespace. There is a lot of codeblocks that are bunched up. Adding newlines will make it easier to look at and read.

File: xbmc/utils/MovingSpeed.h
Comment (CastagnaIT) [reply xbmc/utils/MovingSpeed.h:19]: i will need this in with a followup PR for some XML controls
but i can remove it and add it in the followup PR

File: xbmc/video/PlayerController.cpp
Comment (CastagnaIT) [reply xbmc/video/PlayerController.cpp:522]: here it is intended to avoid decimals

File: xbmc/utils/MovingSpeed.h
Comment (enen92) [reply xbmc/utils/MovingSpeed.h:36]: Better use the initialization with {}?

File: xbmc/video/PlayerController.cpp
Comment (lrusak) [reply xbmc/video/PlayerController.cpp:522]: Yep, and you can avoid the cast

File: xbmc/utils/MovingSpeed.h
Comment (lrusak) [reply xbmc/utils/MovingSpeed.h:19]: Please do 😊

File: xbmc/video/PlayerController.cpp
Comment (lrusak) [reply xbmc/video/PlayerController.cpp]: Just an FYI after testing this it seems fmt will round the value rather than truncate. It's probably fine but worth pointing out: https://godbolt.org/z/Ec8xrscej

basically
```c++

  float value = 1234.5678;

  fmt::print("float: {:.0f}\n", value);
  fmt::print("int  : {}\n", static_cast<int>(value));
```

output:
```
float: 1235
int  : 1234
```

Comment (CastagnaIT) [reply xbmc/video/PlayerController.cpp]: I think it is ok, this value has differents conversions/scale/changes from us and also in the libass library, i don't think it matters much it's just for general user reference

**Comment by enen92** [comment]
jenkins build this please

**Comment by CastagnaIT** [comment]
@lrusak just to know have you to do a second review?

**Comment by Tourosi** [comment]
Before merge, could you correct an expression? A popup for shifting subtitles during playback says "Top-left overscan compensation". I think it should be "Subtitles position" or some expression like that. Thank you.

**Comment by CastagnaIT** [comment]
@Tourosi are you sure to have tested this PR in right way? because i see the right text displayed

File: xbmc/cores/VideoSettings.h
Comment (garbear) [inline xbmc/cores/VideoSettings.h:104]: As long as we're adding new members, they should be initialized in the header:

```c++
  int m_subtitleVerticalPosition{0};
  bool m_subtitleVerticalPositionSave{false};
```

File: xbmc/interfaces/builtins/PlayerBuiltins.cpp
Comment (garbear) [inline xbmc/interfaces/builtins/PlayerBuiltins.cpp:690]: You'll notice that an existing line is changed, dirtying git history. This can be avoided by using a trailing comma (which C++ allows for precisely for this reason)

File: xbmc/settings/windows/GUIWindowSettingsScreenCalibration.cpp
Comment (garbear) [inline xbmc/settings/windows/GUIWindowSettingsScreenCalibration.cpp:510]: I nullptr check should follow immediately after a `dynamic_cast` in this case

File: xbmc/settings/windows/GUIWindowSettingsScreenCalibration.cpp
Comment (garbear) [inline xbmc/settings/windows/GUIWindowSettingsScreenCalibration.cpp:527]: `const auto&`

File: xbmc/settings/windows/GUIWindowSettingsScreenCalibration.cpp
Comment (garbear) [inline xbmc/settings/windows/GUIWindowSettingsScreenCalibration.cpp:528]: This is more style preference, but I prefer `const float` for intermediate variables that aren't expected to change. Newer language like rust embody this philosophy by default-const.

File: xbmc/utils/MovingSpeed.h
Comment (garbear) [inline xbmc/utils/MovingSpeed.h:42]: I like that you encapsulated the moving speed in this new class. The class could use a really top-quality doxy string that scopes what use case the new class is solving, and some description of the desired motion effect.

File: xbmc/interfaces/builtins/PlayerBuiltins.cpp
Comment (CastagnaIT) [reply xbmc/interfaces/builtins/PlayerBuiltins.cpp:690]: I did not understand what you are proposing here please provide example

Comment (garbear) [reply xbmc/interfaces/builtins/PlayerBuiltins.cpp:690]: Change:

```c++
           {"subtitleshiftdown",   {"Shift down the subtitle position", 0, SubtitleShiftDown}}
```

to:

```c++
           {"subtitleshiftdown",   {"Shift down the subtitle position", 0, SubtitleShiftDown}},
```

**Comment by garbear** [comment]
Sorry to come in so late with comments but everything is super minor. In general I approve this PR.

File: xbmc/utils/MovingSpeed.h
Comment (CastagnaIT) [reply xbmc/utils/MovingSpeed.h:21]: i think i can remove it by add a derived private struct like this

```c++
class CMovingSpeed
{
...
private:
  struct EventCfgInternal : EventCfg{
    float m_currentVelocity = 1.0f;
    uint32_t m_lastFrameTime = 0;
  }
}
```

better?

Comment (garbear) [reply xbmc/utils/MovingSpeed.h:21]: I see less encapsulation breakage so I'd say so

File: xbmc/cores/VideoSettings.h
Comment (CastagnaIT) [reply xbmc/cores/VideoSettings.h]: done

File: xbmc/utils/MovingSpeed.h
Comment (CastagnaIT) [reply xbmc/utils/MovingSpeed.h:42]: add it i hope no make writing mistakes

File: xbmc/utils/MovingSpeed.h
Comment (CastagnaIT) [reply xbmc/utils/MovingSpeed.h]: add a private struct

File: xbmc/utils/MovingSpeed.h
Comment (garbear) [reply xbmc/utils/MovingSpeed.h:42]: looks great!

**Review by garbear — APPROVED** [review]
Gotta say, this is a model PR. Great work!

File: xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.h
Comment (howie-f) [inline xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.h:147]: this method can be `const` - not a blocker

Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.h]: done

Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.h]: re-done :)

**Comment by CastagnaIT** [comment]
I don't want to seem nagging but this is last call
if other reviews are needed to those already made
i want to merge within this week

**Review by enen92 — APPROVED** [review]

**Comment by monsoons** [comment]
Great Job! I love the new (save) feature.

It worked on files, but on Netflix addon, the subtitles is fixed below and doesn't shift.
Am I missing some settings?


The two ways of positioning subtitles permanently work differently when Kodi restarts.
If you quit Kodi and restart Kodi, the last position value is not saved when using the new builtin commands SubtitleShiftUp/Down(save). I have to use the calibration window menu to save permanently despite the restart. I wonder if that is the way it is intended.

I tried to edit the value on the file "userdata/guisettings.xml" <subtitles>value</subtitles>. But this way doesn't seem to work any more.

**Comment by CastagnaIT** [comment]
N@tflix use webvtt without override settings enabled you cannot shift subs due to his metadata

> SubtitleShiftUp/Down(save). I have to use the calibration window menu to save permanently despite the restart. I wonder if that is the way it is intended.

ATM i dont remember i think dont implemented saving changes to file for a reason,
i will take a second look to better understand what can be done

**Comment by monsoons** [comment]
> N@tflix use webvtt without override settings enabled you cannot shift subs due to his metadata


Even though I set override subtitles styles to positions, vtt subtitles are not shifted on latest build, whereas ASS subtitles are shifted. I could shift vtt subtitles on previous released builds. (tested on Windows 10 PC , android devices)

**Comment by Dis90** [comment]
I noticed this problem which probably is related to this PR. At first subtitles are in correct place (bottom of screen) but after finishing first episode and second episode starts subtitles jumps to top of screen. I'm watching local mkv files with srt subtitles.

I can open new issue if you want.

**Comment by howie-f** [comment]
yes, i can confirm this @Dis90, maybe we don‘t need a new issue. let‘s hear what @CastagnaIT thinks

**Comment by CastagnaIT** [comment]
I have pushed fixes for problems:
- webvtt override position setting not working (to allow shift subs)
- subs wrong position after sequential video playback

instead to make saving permanently position with SubtitleShiftUp/Down(save) so value kept after reboot,
the problem appears more complex and ATM i have not found a good solution

**Pull request merged**
